pkg/loop/internal/plugin_service.go

```diff
@@ -212,11 +209,47 @@ func (s *pluginService[P, S]) closeClient() (err error) {
 	return
 }
 
-func (s *pluginService[P, S]) wait(ctx context.Context) error {
+func (s *PluginService[P, S]) WaitCtx(ctx context.Context) error {
 	select {
 	case <-ctx.Done():
 		return context.Cause(ctx)
 	case <-s.serviceCh:
 		return nil
+	case <-s.stopCh:
+		return errors.New("service was stopped while waiting")
```

Comment:
I wonder if it matters that this will no longer satisfy `errors.Is(err, context.Canceled)`. Might want to wrap it :shrug: